### PR TITLE
CVS-54174 model layout

### PR DESCRIPTION
**Pull request by dkalinowski**

### CODE REVIEW

File: example_client/faces_analysis_pipeline_client.py [hunk 1]
```diff
@@ -32,7 +32,8 @@
 parser.add_argument('--face_images_save_path', required=False, default='', help='If specified, face images will be saved to disk.')
 parser.add_argument('--image_width', required=False, default=600, help='Original image width. default: 600')
 parser.add_argument('--image_height', required=False, default=400, help='Original image height. default: 400')
-parser.add_argument('--image_layout', required=False, default='CHW', choices=['CHW', 'HWC'], help='Original image layout. default: CHW')
+parser.add_argument('--input_image_layout', required=False, default='CHW', choices=['CHW', 'HWC'], help='Original image layout. default: CHW')
```
Comment (dtrawins) [inline example_client/faces_analysis_pipeline_client.py:35]: help sentence is a bit missleading. what is the original? maybe model input layout would be clearer?

File: example_client/faces_analysis_pipeline_client.py [hunk 2]
```diff
@@ -32,7 +32,8 @@
...
+parser.add_argument('--output_image_layout', required=False, default='CHW', choices=['CHW', 'HWC'], help='Target image layout. default: CHW')
```
Comment (dtrawins) [inline example_client/faces_analysis_pipeline_client.py:36]: Model output layout defined in the ovms response?

File: src/binaryutils.cpp [hunk 3]
```diff
@@ -131,14 +106,9 @@ Status resizeMat(const cv::Mat& src, cv::Mat& dst, const std::shared_ptr<TensorI
...
-        if ((unsigned int)(input.channels()) != tensorInfo->getShape()[1]) {
-            SPDLOG_DEBUG("Binary sent to input: {} has invalid number of channels. Expected: {} Actual: {}", tensorInfo->getMappedName(), tensorInfo->getShape()[1], input.channels());
-            return StatusCode::INVALID_NO_OF_CHANNELS;
-        }
-    } else if (tensorInfo->getLayout() == InferenceEngine::Layout::NHWC) {
-        if ((unsigned int)(input.channels()) != tensorInfo->getShape()[3]) {
-            SPDLOG_DEBUG("Binary sent to input: {} has invalid number of channels. Expected: {} Actual: {}", tensorInfo->getMappedName(), tensorInfo->getShape()[3], input.channels());
+    if (tensorInfo->getLayout() == InferenceEngine::Layout::NHWC) {
+        if ((unsigned int)(input.channels()) != tensorInfo->getEffectiveShape()[3]) {
+            SPDLOG_DEBUG("Binary sent to input: {} has invalid number of channels. Expected: {} Actual: {}", tensorInfo->getMappedName(), tensorInfo->getEffectiveShape()[3], input.channels());
```
Comment (dtrawins) [inline src/binaryutils.cpp:111]: Binary data

File: src/custom_nodes/model_zoo_intel_object_detection/README.md [hunk 4]
```diff
@@ -55,6 +55,8 @@ Parameters can be defined in pipeline definition in OVMS configuration file. [Re
 | original_image_height  | Required input image height |  | &check; |
 | target_image_width | Target width of the boxes in output. Boxes in the original image will be resized to that value.  |  | &check; |
 | target_image_height  | Target width of the boxes in output. Boxes in the original image will be resized to that value. |  | &check; |
+| original_image_layout | Defines original image layout | NCHW | |
```
Comment (dtrawins) [inline src/custom_nodes/model_zoo_intel_object_detection/README.md:58]: input image layout?

File: src/custom_nodes/model_zoo_intel_object_detection/README.md [hunk 5]
```diff
@@ -55,6 +55,8 @@ Parameters can be defined in pipeline definition in OVMS configuration file. [Re
...
+| target_image_layout | Defines target image layout | NCHW | |
```
Comment (dtrawins) [inline src/custom_nodes/model_zoo_intel_object_detection/README.md:59]: output image layout?

File: src/gathernodeinputhandler.cpp [hunk 6]
```diff
@@ -68,10 +68,12 @@ Status GatherNodeInputHandler::notifyFinishedDependency() {
     }
     for (auto& [inputName, shardMap] : shardsStorage) {
         const auto shardsCount = shardMap.size();
-        SPDLOG_LOGGER_DEBUG(dag_executor_logger, "Consolidating: {} shards for input: {}", shardsCount, inputName);
+        SPDLOG_LOGGER_INFO(dag_executor_logger, "Consolidating: {} shards for input: {}", shardsCount, inputName);
         session_id_t firstShardId = 0;
         auto firstShardTensorDesc = shardMap.at(firstShardId)->getTensorDesc();
-        auto shardDims = firstShardTensorDesc.getDims();
+        // auto shardDims = firstShardTensorDesc.getDims();
+        std::cout << TensorInfo::shapeToString(firstShardTensorDesc.getDims()) << " " << TensorInfo::shapeToString(firstShardTensorDesc.getBlockingDesc().getBlockDims()) << std::endl;
```
Comment (dtrawins) [inline src/gathernodeinputhandler.cpp:75]: is it to be removed?

File: example_client/faces_analysis_pipeline_client.py (hunk 1 above)
Comment (dkalinowski) [reply example_client/faces_analysis_pipeline_client.py:35]: This client is removed on main branch and will be removed from develop anyway.

File: example_client/faces_analysis_pipeline_client.py (hunk 2 above)
Comment (dkalinowski) [reply example_client/faces_analysis_pipeline_client.py:36]: This client is removed on main branch and will be removed from develop anyway.

Comment (dkalinowski) [reply example_client/faces_analysis_pipeline_client.py:36]: Transformed detected images layout

File: example_client/faces_analysis_pipeline_client.py (hunk 1 above)
Comment (dkalinowski) [reply example_client/faces_analysis_pipeline_client.py:35]: Pipeline input image layout

Comment (dkalinowski) [reply example_client/faces_analysis_pipeline_client.py:35]: CHW -> NHWC

File: src/binaryutils.cpp [hunk 7]
```diff
@@ -226,14 +195,8 @@ InferenceEngine::Blob::Ptr createBlobFromMats(const std::vector<cv::Mat>& images
...
     for (cv::Mat image : images) {
-        if (tensorInfo->getLayout() == InferenceEngine::Layout::NCHW) {
-            auto imgBuffer = reorder_to_nchw((T*)image.data, image.rows, image.cols, image.channels());
-            memcpy(ptr + offset, (char*)imgBuffer.data(), image.total() * image.elemSize());
-            offset += (image.total() * image.elemSize());
-        } else {
-            memcpy(ptr + offset, (char*)image.data, image.total() * image.elemSize());
-            offset += (image.total() * image.elemSize());
-        }
+        memcpy(ptr + offset, (char*)image.data, image.total() * image.elemSize());
```
Comment (dkalinowski) [inline src/binaryutils.cpp:198]: Instead of incrementing offset increment ptr to avoid overflow

**Review by michalkulakowski — APPROVED** [review]

File: src/custom_nodes/model_zoo_intel_object_detection/README.md [hunk 8]
```diff
@@ -55,8 +55,8 @@ Parameters can be defined in pipeline definition in OVMS configuration file. [Re
 | original_image_height  | Required input image height |  | &check; |
 | target_image_width | Target width of the boxes in output. Boxes in the original image will be resized to that value.  |  | &check; |
 | target_image_height  | Target width of the boxes in output. Boxes in the original image will be resized to that value. |  | &check; |
-| original_image_layout | Defines original image layout | NCHW | |
-| target_image_layout | Defines target image layout | NCHW | |
+| original_image_layout | Defines input image image layout | NCHW | |
```
Comment (dtrawins) [inline src/custom_nodes/model_zoo_intel_object_detection/README.md:58]: double image

File: src/custom_nodes/model_zoo_intel_object_detection/README.md [hunk 9]
```diff
@@ -55,6 +55,8 @@ Parameters can be defined in pipeline definition in OVMS configuration file. [Re
 | original_image_height  | Required input image height |  | &check; |
 | target_image_width | Target width of the boxes in output. Boxes in the original image will be resized to that value.  |  | &check; |
 | target_image_height  | Target width of the boxes in output. Boxes in the original image will be resized to that value. |  | &check; |
+| original_image_layout | Defines input image image layout | NCHW | |
+| target_image_layout | Defines output images layout | NCHW | |
```
Comment (dtrawins) [inline src/custom_nodes/model_zoo_intel_object_detection/README.md:59]: Defines the data layout of detected object images in the node output

**Review by dtrawins — APPROVED** [review]

**Pull request merged**